Source/WebCore/rendering/RenderListMarker.cpp

```diff
@@ -1840,16 +1837,29 @@ void RenderListMarker::updateContent()
         return;
     }
 
-    auto type = style().listStyleType();
-    switch (type) {
-    // FIXME: handle CSSCounterStyle case rdar://102988393.
-    case ListStyleType::CustomCounterStyle:
-    case ListStyleType::String:
+    auto markerAsString = [&] {
         m_textWithSuffix = style().listStyleStringValue();
         m_textWithoutSuffixLength = m_textWithSuffix.length();
-        // FIXME: Depending on the string value, we may need the real bidi algorithm.
+        // FIXME: Depending on the string value, we may need the real bidi algorithm. (rdar://106139180)
         m_textIsLeftToRightDirection = u_charDirection(m_textWithSuffix[0]) != U_RIGHT_TO_LEFT;
+    };
+
+    auto type = style().listStyleType();
+    switch (type) {
+    case ListStyleType::String:
+        markerAsString();
         break;
+    case ListStyleType::CustomCounterStyle: {
+        auto* counter = counterStyle();
+        if (counter) {
+            auto text = makeString(counter->prefix(), counter->text(m_listItem->value()));
+            m_textWithSuffix = makeString(text, counter->suffix());
+            m_textWithoutSuffixLength = text.length();
+            m_textIsLeftToRightDirection = u_charDirection(text[0]) != U_RIGHT_TO_LEFT;
+        } else
+            markerAsString();
```

Comment:
This seems incorrect? If the counter-style doesn't exist, it should fallback to decimal I think.